src/xdplwf/generic.c

```diff
@@ -264,19 +349,94 @@ XdpGenericDereferenceDatapath(
 
             FRE_ASSERT(!CancelledTimer);
             FRE_ASSERT(StartedTimer);
-            return;
+            goto Exit;
         }
 
         KeClearEvent(&Datapath->ReadyEvent);
-        *NeedRestart = TRUE;
+        NeedRestart = TRUE;
     }
 
     FRE_ASSERT(Datapath->ReferenceCount > 0);
     --Datapath->ReferenceCount;
 
-    if (*NeedRestart) {
+    if (NeedRestart) {
         XdpGenericDereference(Generic);
     }
+
+Exit:
+
+    return NeedRestart;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+VOID
+XdpGenericAttachDatapath(
+    _In_ XDP_LWF_GENERIC *Generic,
+    _In_ BOOLEAN RxDatapath,
+    _In_ BOOLEAN TxDatapath
+    )
+{
+    BOOLEAN NeedRestart = FALSE;
+    LARGE_INTEGER Timeout;
+    VOID *ReadyEvents[2];
+    UINT32 ReadyEventCount = 0;
+
+    ASSERT(RxDatapath || TxDatapath);
+
+    RtlAcquirePushLockExclusive(&Generic->Lock);
+    if (RxDatapath) {
+        NeedRestart |= XdpGenericReferenceDatapath(Generic, &Generic->Rx.Datapath);
+        ReadyEvents[ReadyEventCount++] = &Generic->Rx.Datapath.ReadyEvent;
+    }
+    if (TxDatapath) {
+        NeedRestart |= XdpGenericReferenceDatapath(Generic, &Generic->Tx.Datapath);
+        ReadyEvents[ReadyEventCount++] = &Generic->Tx.Datapath.ReadyEvent;
+    }
+    RtlReleasePushLockExclusive(&Generic->Lock);
+
+    C_ASSERT(RTL_NUMBER_OF(ReadyEvents) <= THREAD_WAIT_OBJECTS);
+    ASSERT(ReadyEventCount > 0 && ReadyEventCount <= RTL_NUMBER_OF(ReadyEvents));
+
+    if (NeedRestart) {
+        TraceVerbose(
+            TRACE_GENERIC, "IfIndex=%u Requesting datapath attach RX=%u TX=%u",
+            Generic->IfIndex, RxDatapath, TxDatapath);
+        XdpGenericRequestRestart(Generic);
+    }
+
+    Timeout.QuadPart =
+        -1 * RTL_MILLISEC_TO_100NANOSEC(GENERIC_DATAPATH_RESTART_TIMEOUT_MS);
+    KeWaitForMultipleObjects(
+        ReadyEventCount, ReadyEvents, WaitAll, Executive, KernelMode, FALSE, &Timeout, NULL);
```

Comment:
What if it times out?

Reply:
This has always been a best-effort attempt to cause XDP objects (AF_XDP sockets, programs) to be created/bound only after the NDIS datapath is unpaused. If the NDIS data path isn't ready within this timeout, we just go ahead anyways, since nothing terrible will happen. The program/socket will just be blackholed until NDIS finally gets around to unpausing, if it ever does.

Reply:
Basically, we're trying to eliminate an annoying race condition for apps (I created a socket, and it can't send!) without introducing a hard dependency on NDIS timeliness, which is never guaranteed.

Reply:
Do you think it worth having a breadcrumb (ETW event) to warn of this happening?

Reply:
We have logs tracking all NDIS pause/restarts. I don't think it would be harmful to have ETW, but we currently have a precedent to use WPP for the control path. IMO we should keep the existing style of WPP logs and convert to something more modern, e.g. tracelogging.